oximeter/instruments/src/kstat/mod.rs

```diff
@@ -0,0 +1,218 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types for publishing kernel statistics via oximeter.
+
+use chrono::DateTime;
+use chrono::Utc;
+use kstat_rs::Data;
+use kstat_rs::Error as KstatError;
+use kstat_rs::Kstat;
+use kstat_rs::NamedData;
+use kstat_rs::NamedType;
+use oximeter::FieldValue;
+use oximeter::MetricsError;
+use oximeter::Sample;
+use oximeter::Target;
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::time::Duration;
+
+pub mod link;
```

Comment:
It's a good question, and I'm not sure. I agree that it's unfortunate to link in stuff around datalinks to Propolis if it only cares about the sampler type. OTOH, I had originally intended for this crate to actually have the `Target` and `Metric` definitions for broadly-useful statistics. E.g., the `http` module has those for tracking request latency.

The solution in that case is to put the `http` stuff behind a feature flag. Would something similar for links be amenable? Specifically, I can make a `datalink` feature, distinct from the `kstat` feature. The sled agent would use both, and Propolis could decide to just use the `kstat` feature and define its `Target`s and `Metrics` locally.

Reply:
Actually, I've a question about this. I realize I've been assuming that the sled agent would produce all metrics related to kstats. That was partially based on an assumption that the kstats would not be visible inside the zone, but that's not correct. `kstat -l link:::` for example, shows the relevant data link inside the zones just fine.

But, it does seem a bit confusing to produce samples from kstats for memory or vCPU usage from inside the zone, while the link stats are produced from the GZ. I don't really have a strong opinion about where that should happen, actually, but I imagine we'll want to do it from the same place.

Any thoughts on that?

Reply:
I thought about it some more, and I don't think I have a concrete example to hand of a kstat that we'd grab from anywhere other than sled agent (and I have to admit that my Propolis-as-consumer example is FUD and not something we're concretely planning). So I'm having trouble feeling too strongly one way or the other, and so would vote for leaving this as-is for now so that we can keep the wheels turning. We can always refactor later if we find a strong reason to do so.